### DISCUSSION TITLE
please allow disabling implicit conversions in non-single-file use

### DISCUSSION THREAD

**Discussion by maddanio** [nlohmann/json#2622] [Q&A] [answered]
Hello,

I recently had to disable implicit conversions as they finally completely broke for us: If one of the types that json converts to are output streamable then streaming out json will immediately become ambiguous. Maybe that is further incentive for deprecating/disablong this and/or defaulting to disabled implicit conversions.

The problem is the disabling of implicit conversions seemingly only has been done for the single file header, but we really like to use json_fwd to speed up compilation, which is not possible with single file. Is it possible to enable those for non-single-file?

**Comment by maddanio** [comment]
oh, I thought you had just deleted this as of my me being too stupid to see that somehow misteriously I completely messed up, and of course the implicit switch works also on multi-file. the only thing to "discuss" here is maybe deprecating them as they can very easily go wild? anyway, sorry again.

**Reply by nlohmann** [reply]
I moved the issue, because it is not a bug - you can disable implicit conversions, see https://github.com/nlohmann/json/discussions/2622#discussioncomment-338500. Unfortunately, the message I wrote was not sent...

**Comment by gregmarr** [comment, marked as answer]
The single header is a combination of the individual headers, so anything that works for the single headers should work for the multiple headers.  Can you provide more detail about what is not working for you?

**Reply by maddanio** [reply]
Ah sorry, thought i had made it clear it was my mistake. Btw, is there a chance to make the initializer list constructor optional as well? It keeps on making single element lists...

**Reply by gregmarr** [reply]
I thought that might be the case, but wasn't sure.  The last sentence about deprecating them was the really confusing part.  

> Btw, is there a chance to make the initializer list constructor optional as well? It keeps on making single element lists...

You can always create a PR for that, not sure if @nlohmann would accept it or not. I know that was a big feature at one point.  Not sure if the issues with it make it worth removing.

**Reply by nlohmann** [reply]
- Changing the defaults will be a breaking change and will not happen before 3.0.0.
- An MR for the initializer list would be welcomed - but it would need to be a solution similar to the implicit conversions, namely a macro to selectively switch off that constructor.
- If apart from this you have suggestions how to improve documentation, please let me know.
